samples/wifi/shell/src/wifi_cred_shell.c

```diff
@@ -0,0 +1,229 @@
+		.header.ssid_len = strlen(argv[1]),
+		.header.type = parse_sec_type(argv[2]),
+	};
+
+	memcpy(creds.header.ssid, argv[1], creds.header.ssid_len);
+
+	if (creds.header.type == WIFI_SECURITY_TYPE_UNKNOWN) {
+		printk("cannot parse security type!\n");
+		goto help;
+	}
+
+	size_t arg_idx = 3;
+
+	if (creds.header.type == WIFI_SECURITY_TYPE_PSK ||
+	    creds.header.type == WIFI_SECURITY_TYPE_PSK_SHA256 ||
+	    creds.header.type == WIFI_SECURITY_TYPE_SAE) {
+		/* parse passphrase */
+		if (argc < 4) {
+			printk("missing password\n");
+			goto help;
+		}
+		creds.password_len = strlen(argv[3]);
+		if (creds.password_len > CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH ||
+			(
+				(creds.header.type == WIFI_SECURITY_TYPE_PSK ||
+				 creds.header.type == WIFI_SECURITY_TYPE_PSK_SHA256) &&
+				creds.password_len > WIFI_PSK_MAX_LEN
+			)
+		) {
+			printk("psk is too long for this security type!\n");
+			goto help;
+		}
+		memcpy(creds.password, argv[3], creds.password_len);
+		++arg_idx;
+	}
+
+	if (arg_idx < argc) {
+		/* look for bssid */
+		ret = sscanf(argv[arg_idx], MACSTR,
+			     &creds.header.bssid[0], &creds.header.bssid[1],
+			     &creds.header.bssid[2], &creds.header.bssid[3],
+			     &creds.header.bssid[4], &creds.header.bssid[5]);
+		if (ret == 6) {
+			creds.header.flags |= WIFI_CREDENTIALS_FLAG_BSSID;
+			++arg_idx;
+		}
+	}
+
+	if (arg_idx < argc) {
+		/* look for band */
+		enum wifi_frequency_bands band = parse_band(argv[arg_idx]);
+
+		if (band == WIFI_FREQ_BAND_2_4_GHZ) {
+			creds.header.flags |= WIFI_CREDENTIALS_FLAG_2_4GHz;
+			++arg_idx;
+		}
+		if (band == WIFI_FREQ_BAND_5_GHZ) {
+			creds.header.flags |= WIFI_CREDENTIALS_FLAG_5GHz;
+			++arg_idx;
+		}
+	}
+
+	if (arg_idx < argc) {
+		/* look for favorite flag */
+		if (strcmp("favorite", argv[arg_idx]) == 0) {
+			creds.header.flags |= WIFI_CREDENTIALS_FLAG_FAVORITE;
+			++arg_idx;
+		}
+	}
+
+	if (arg_idx != argc) {
+		for (size_t i = arg_idx; i < argc; ++i) {
+			printk("unparsed arg: [%s]\n", argv[i]);
+		}
+	}
+
+	return wifi_credentials_set_personal_struct(&creds);
+help:
+	printk("usage: wifi_cred add \"network name\""
+		       " {OPEN, WPA2-PSK, WPA2-PSK-SHA256, WPA3-SAE}"
+		       " [psk]"
+		       " [bssid]"
+		       " [{2.4GHz, 5GHz}]"
+		       " [favorite]\n");
+	return -EINVAL;
+}
+
+static int cmd_delete_network(const struct shell *sh, size_t argc, char *argv[])
+{
+	if (argc != 2) {
+		printk("usage: wifi_cred delete \"network name\"\n");
+		return -EINVAL;
+	}
+	printk("  deleting network ssid: \"%s\", ssid_len: %d\n", argv[1], strlen(argv[1]));
+	return wifi_credentials_delete_by_ssid(argv[1], strlen(argv[1]));
+}
+
+static int cmd_list_networks(const struct shell *sh, size_t argc, char *argv[])
+{
+	wifi_credentials_for_each_ssid(print_network_info_helper, NULL);
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_wifi_cred,
+	SHELL_CMD(auto_connect,	NULL, "Connect to any stored network.",	cmd_auto_connect),
```

Comment:
Perhaps we should modify the original connect command to support connecting to configured networks if no params are provided rather than introduce a new one?

Reply:
This would be like the initial proposal I made. The blocker for this approach is [here](https://github.com/nrfconnect/sdk-zephyr/blob/eb7519d36a0c2819d5d52328751395a9f0234880/subsys/net/l2/wifi/wifi_mgmt.c#L42).
The `params` object is checked inside Zephyr, so we cannot really mess with it much.

Reply:
I think that if we are thinking to ever upstream the `wpa_supplicant` we should avoid adding any dependencies to NCS modules in there. It should only depend on a Wi-Fi driver through a generic API.

Reply:
One question I have is: is it true that anyone using `wpa_supplicant` will face the problem of persistent credentials/network storage? I also wonder how this is in fact done for the linux port of wpa_supplicant? AFAIR there is a configuration file for the supplicant. How are the contents of that file propagated to the supplicant?

The point is: if we think about upstreaming the supplicant the extra modules like this one and the credentials storage should probably go upstream too.